docs/src/docs/asciidoc/developer/clustered.adoc

```diff
@@ -0,0 +1,175 @@
+state stored within the Voltron server.  First I will outline the basics of the
+proposal, then I will give some examples of its usage to implement a variety of
+single key operations on the cache.  Finally I will outline some implementation
+decisions we may face around performance tradeoffs with the proposal.
+
+== The Proposal
+
+Ehcache supports an open type system on its keys.  This means user can use any
+'serializable' type as a key, this means the server cannot make any judgements
+about it's ability to perform equals comparisons on keys (or values in the case
+of concurrent operations).  Therefore to support equality comparison on clients
+without performing CAS loops or taking exclusive locks across a network this
+scheme has been proposed.
+
+== Server Side Data Structure
+
+The server will store cache data as a multi-map where the key is the cache-key
+hash (minimally just the Java hashcode, but potentially something wider than
+that), and the value is a sequence of binary payloads:
+
+["ditaa", "server-data-structure", "png"]
+--------------------------------------------------------------------------------
+             +----+    +----------------------------------------------+
+             |    |    | +-----------+ +-----------+ +-----------+    |
+             |hash| => | |binary blob| |binary blob| |binary blob| ...|
+             |    |    | +-----------+ +-----------+ +-----------+    |
+             +----+    +----------------------------------------------+
+
+             +----+    +----------------------------------------------+
+             |    |    | +-----------+ +-----------+ +-----------+    |
+             |hash| => | |binary blob| |binary blob| |binary blob| ...|
+             |    |    | +-----------+ +-----------+ +-----------+    |
+             +----+    +----------------------------------------------+
+
+             +----+    +----------------------------------------------+
+             |    |    | +-----------+ +-----------+ +-----------+    |
+             |hash| => | |binary blob| |binary blob| |binary blob| ...|
+             |    |    | +-----------+ +-----------+ +-----------+    |
+             +----+    +----------------------------------------------+
+--------------------------------------------------------------------------------
+
+The server supports the following operations:
+
+ * `List<binary-blob> get(hash)`: returns the list for the given hash
+ * `void append(hash, binary-blob)`: adds the given binary blob to the end of
+   the list for the given hash
+ * `list<binary-blob> getAndAppend(hash, binary-blob)`: the same as append but
+   also returns the list immediately prior to mutation (atomically).
+ * `void replace(hash, list<binary-blob>, list<binary-blob>)`: replaces the
+   first occurrence of the given subsequence with the supplied sequence
+
+This is the entirety of the functionality and data structure exposed by the
```

Comment:
'exposed by the client'  - I know this is literally true, but "client" can be a matter of perspective, so it may be worth noting to readers that it is also the api exposed *to* the client side to work with / upon.